### PR TITLE
added multi-prototype head for SWav

### PR DESCRIPTION
**Pull request by Atharva-Phatak**
This PR implements a multi-prototype head for SWav as discussed in #944. 

Looking forward to feedback.

### CODE REVIEW

**Comment by Atharva-Phatak** [comment]
Perfect the tests are working. Looking forward to the feedback and code reviews

File: lightly/models/modules/heads.py [hunk 1]
```diff
@@ -370,14 +370,20 @@ class SwaVPrototypes(ProjectionHead):
...
+        #Default to a list of 1 if n_prototypes is an int.
+        self.n_prototypes = n_prototypes if isinstance(n_prototypes, list) else [n_prototypes]
+        for layerNum, k in enumerate(self.n_prototypes):
+            self.add_module("prototypes" + str(layerNum), nn.Linear(input_dim, k, bias = False))
+
+    def forward(self, x):
+        out = []
+        for head in range(self.number_of_heads):
+            out.append(getattr(self, "prototypes" + str(head))(x))
+        return out
```
Comment (guarin) [inline lightly/models/modules/heads.py:386]: This changes the behaviour of  the `SwaVPrototypes` head and might break existing code. Before this change it returned a single tensor and now it returns a list of tensors.
 
I would propose to keep the old behaviour if users specify `n_prototypes` with a single number. If `n_prototypes` is a list, the forward function should also return a list. This could be implemented by adding a `self.single_head = isinstance(n_prototypes, int)` line in the `__init__` method and checking the property in the forward function.

File: lightly/models/modules/heads.py [hunk 2]
```diff
@@ -370,14 +370,20 @@ class SwaVPrototypes(ProjectionHead):
...
     """
     def __init__(self,
-                 input_dim: int = 128,
-                 n_prototypes: int = 3000):
-        super(SwaVPrototypes, self).__init__([])
-        self.layers = nn.Linear(input_dim, n_prototypes, bias=False)
-
-    def normalize(self):
-        """Normalizes the prototypes so that they are on the unit sphere."""
-        utils.normalize_weight(self.layers.weight)
```
Comment (guarin) [inline lightly/models/modules/heads.py:380]: I don't think this should be deleted. Instead we should loop over all heads and normalise them.

**Review by guarin — CHANGES_REQUESTED** [review]
Thanks a lot for the PR 💯 

The code looks great. I added some comments and suggestions, mostly regarding backwards compatibility.

File: lightly/models/modules/heads.py [hunk 3]
```diff
@@ -370,14 +370,20 @@ class SwaVPrototypes(ProjectionHead):
...
+                input_dim: int = 128,
+                n_prototypes: Union[List[int], int] = 3000):
+        super(SwaVPrototypes, self).__init__()
+        self.number_of_heads = len(n_prototypes) if isinstance(n_prototypes, list) else 1
+        #Default to a list of 1 if n_prototypes is an int.
+        self.n_prototypes = n_prototypes if isinstance(n_prototypes, list) else [n_prototypes]
+        for layerNum, k in enumerate(self.n_prototypes):
+            self.add_module("prototypes" + str(layerNum), nn.Linear(input_dim, k, bias = False))
```
[the remark this answers is not in the document]
Comment (guarin) [reply lightly/models/modules/heads.py:380]: I think we should use the builtin functionality provided by pytorch to register new modules.

File: lightly/models/modules/heads.py (hunk 1 above)
Comment (Atharva-Phatak) [reply lightly/models/modules/heads.py:386]: Hey @guarin. So just to clarify, the output should be a tensor if n_prototypes is an int else, return a list. I think this should be easy to implement. Am I correct ?

Comment (guarin) [reply lightly/models/modules/heads.py]: Thanks for implementing the feature :)

File: lightly/models/modules/heads.py (hunk 2 above)
Comment (guarin) [reply lightly/models/modules/heads.py:380]: I think this is the only thing that is missing then we should be good to merge :)

**Review by guarin — COMMENTED** [review]
Thanks a lot for updating the PR. The only thing that is missing is the normalisation, then it should be ready to merge :)

**Comment by Atharva-Phatak** [comment]
Added the normalisation.

**Review by guarin — APPROVED** [review]
Great, thanks a lot 💯
